src/compile/model.ts

```diff
@@ -268,7 +270,14 @@ export abstract class Model {
   public abstract assembleSelectionSignals(): any[];
 
   public abstract assembleSelectionData(data: VgData[]): VgData[];
-  public abstract assembleData(): VgData[];
+  public assembleData(): VgData[] {
+     if (!this.parent) {
```

Comment:
If the rule is only assembleData in the root, why don't we just make sure that it is only called from `compile.ts`?